app/models/renewal/SendTheLargestAmountsOfMoney.scala

```diff
@@ -17,47 +17,92 @@
+
+      import TraversableValidators._
+      import utils.MappingUtils.Implicits.RichRule
 
-  implicit val format = Json.format[SendTheLargestAmountsOfMoney]
+      implicit val emptyToNone: String => Option[String] = {
+        case "" => None
+        case s => Some(s)
+      }
+
+      val seqR = {
+        (seqToOptionSeq[String] andThen flattenR[String] andThen cR)
+          .andThen(minLengthR[Seq[Country]](1) withMessage "error.invalid.country")
+          .andThen(maxLengthR[Seq[Country]](3))
+      }
+
+      (__ \ "largestAmountsOfMoney").read(seqR) map SendTheLargestAmountsOfMoney.apply
+    }
+
+  private implicit def write[A]
+  (implicit
+   a: Path => WriteLike[Seq[Country], A]
+  ): Write[SendTheLargestAmountsOfMoney, A] =
+    To[A] { __ =>
+      import play.api.libs.functional.syntax.unlift
+      (__ \ "largestAmountsOfMoney").write[Seq[Country]] contramap unlift(SendTheLargestAmountsOfMoney.unapply)
+    }
 
-  implicit val formRule: Rule[UrlFormEncoded, SendTheLargestAmountsOfMoney] = From[UrlFormEncoded] { __ =>
+  val formR: Rule[UrlFormEncoded, SendTheLargestAmountsOfMoney] = {
     import jto.validation.forms.Rules._
-    import utils.MappingUtils.Implicits._
-        ((__ \ "country_1").read[Country].withMessage("error.required.renewal.country.name") ~
-          (__ \ "country_2").read[Option[Country]] ~
-          (__ \ "country_3").read[Option[Country]]
-          )(SendTheLargestAmountsOfMoney.apply _)
+    implicitly
+  }
+
+  val jsonR: Reads[SendTheLargestAmountsOfMoney] = {
+    import play.api.libs.functional.syntax._
+    import play.api.libs.json._
+    ((__ \ "country_1").read[Country] and
+      (__ \ "country_2").readNullable[Country] and
+      (__ \ "country_3").readNullable[Country]).tupled map {
+      case (a, Some(b), Some(c)) => SendTheLargestAmountsOfMoney(Seq(a, b, c))
+      case (a, Some(b), None) => SendTheLargestAmountsOfMoney(Seq(a, b))
+      case (a, None, None) => SendTheLargestAmountsOfMoney(Seq(a))
+      case (_, _, _) => SendTheLargestAmountsOfMoney(Seq())
     }
+  }
 
-  implicit val formWrites: Write[SendTheLargestAmountsOfMoney, UrlFormEncoded] = Write { countries =>
-      Map(
-        "country_1" -> Seq(countries.country_1.code),
-        "country_2" -> (countries.country_2.toSeq map { _.code }),
-        "country_3" -> (countries.country_3.toSeq map { _.code })
-      )
+  val formW: Write[SendTheLargestAmountsOfMoney, UrlFormEncoded] = {
+    import jto.validation.forms.Writes._
+    import utils.MappingUtils.spm
+    implicitly
+  }
+
+  val jsonW = Writes[SendTheLargestAmountsOfMoney] { lom =>
+    lom.countries match {
+      case Seq(a, b, c) => Json.obj("country_1" -> a, "country_2" -> b, "country_3" -> c)
+      case Seq(a, b) => Json.obj("country_1" -> a, "country_2" -> b)
+      case Seq(a) => Json.obj("country_1" -> a)
+      case _ => Json.obj()
     }
+  }
+}
+
+object SendTheLargestAmountsOfMoney {
+
+  private object Cache extends SendTheLargestAmountsOfMoney0
```

Comment:
Is Cache the right name for this?

Reply:
same as above just followed some other naming, but I could it change to something else if you want